File: scripts/phantomclaw_run_due_automations.py
from __future__ import annotations

import argparse
import json
import subprocess
import sys
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo

PACKAGE_ROOT = Path(__file__).resolve().parent.parent
if str(PACKAGE_ROOT) not in sys.path:
    sys.path.insert(0, str(PACKAGE_ROOT))

from phantomclaw_codex_migration import build_dispatch_bundle, load_registry, validate_registry

DAY_CODES = ["MO", "TU", "WE", "TH", "FR", "SA", "SU"]


def default_state_path() -> Path:
    return Path.home() / ".config" / "phantomclaw" / "automations" / "state.json"


def default_outbox_dir() -> Path:
    return Path.home() / ".config" / "phantomclaw" / "automation-outbox"


def parse_rrule(rrule: str) -> dict[str, str]:
    parts: dict[str, str] = {}
    for part in (rrule or "").split(";"):
        if "=" not in part:
            continue
        key, value = part.split("=", 1)
        parts[key.upper()] = value
    return parts


def int_values(value: str | None, default: list[int]) -> list[int]:
    if not value:
        return default
    values: list[int] = []
    for item in value.split(","):
        try:
            values.append(int(item))
        except ValueError:
            continue
    return values or default


def day_allowed(parts: dict[str, str], now: datetime) -> bool:
    byday = parts.get("BYDAY")
    if not byday:
        return True
    allowed = {day.strip().upper() for day in byday.split(",") if day.strip()}
    return DAY_CODES[now.weekday()] in allowed


def due_occurrence_key(automation: dict, now: datetime) -> str | None:
    parts = parse_rrule(str(automation.get("rrule") or ""))
    freq = parts.get("FREQ", "").upper()
    if not day_allowed(parts, now):
        return None

    minutes = int_values(parts.get("BYMINUTE"), [0])
    if now.minute not in minutes:
        return None

    if freq == "HOURLY":
        interval = int_values(parts.get("INTERVAL"), [1])[0]
        if interval <= 0:
            interval = 1
        if now.hour % interval != 0:
            return None
        return f"{automation['id']}:{now.strftime('%Y-%m-%dT%H')}:{now.minute:02d}"

    if freq in {"WEEKLY", "DAILY"}:
        hours = int_values(parts.get("BYHOUR"), [0])
        if now.hour not in hours:
            return None
        if freq == "WEEKLY":
            return f"{automation['id']}:{DAY_CODES[now.weekday()]}:{now.strftime('%Y-%m-%d')}:{now.hour:02d}:{now.minute:02d}"
        return f"{automation['id']}:{now.strftime('%Y-%m-%d')}:{now.hour:02d}:{now.minute:02d}"

    return None


def load_state(path: Path) -> dict:
    if not path.exists():
        return {"schema_version": "phantomclaw.scheduler-state.v1", "last_occurrences": {}}
    text = path.read_text().strip()
    if not text:
        return {"schema_version": "phantomclaw.scheduler-state.v1", "last_occurrences": {}}
    return json.loads(text)


def write_state(path: Path, state: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(state, indent=2, sort_keys=True) + "\n")


def write_bundle(bundle: dict, outbox_dir: Path) -> Path:
    automation_id = bundle["report"]["automation_id"]
    path = outbox_dir / automation_id / f"{bundle['run']['run_id']}.bundle.json"
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(bundle, indent=2, sort_keys=True) + "\n")
    return path


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Run due migrated automations through PhantomClaw dispatch bundles")
    parser.add_argument("--registry", type=Path)
    parser.add_argument("--state", type=Path, default=default_state_path())
    parser.add_argument("--outbox", type=Path, default=default_outbox_dir())
    parser.add_argument("--now", help="ISO timestamp override for tests")
    parser.add_argument("--sync", action="store_true")
    parser.add_argument("--phantomclaw-cli", default=str(Path.home() / "Documents" / "GitHub" / "phantomclaw-cli" / "dist" / "cli.js"))
    args = parser.parse_args(argv)

    registry = load_registry(args.registry)
    validate_registry(registry)
    tz = ZoneInfo("Europe/Berlin")
    now = datetime.fromisoformat(args.now).astimezone(tz) if args.now else datetime.now(tz)
    state = load_state(args.state)
    last_occurrences = state.setdefault("last_occurrences", {})

    dispatched: list[dict[str, object]] = []
    skipped: list[dict[str, object]] = []
    for automation in registry["automations"]:
        if automation.get("status") != "ACTIVE":
            skipped.append({"id": automation["id"], "reason": "paused"})
            continue
        occurrence_key = due_occurrence_key(automation, now)
        if not occurrence_key:
            skipped.append({"id": automation["id"], "reason": "not_due"})
            continue
        if last_occurrences.get(automation["id"]) == occurrence_key:
            skipped.append({"id": automation["id"], "reason": "already_dispatched", "occurrence": occurrence_key})
            continue

        run_id = f"{automation['id']}-{now.strftime('%Y%m%dT%H%M%S%z')}"
        bundle = build_dispatch_bundle(automation, run_id=run_id)
        bundle_path = write_bundle(bundle, args.outbox)
        last_occurrences[automation["id"]] = occurrence_key
        entry: dict[str, object] = {"id": automation["id"], "bundle": str(bundle_path), "occurrence": occurrence_key}
        if args.sync:
            completed = subprocess.run(
                ["node", args.phantomclaw_cli, "bundle", "sync", str(bundle_path)],
                check=False,
                capture_output=True,
                text=True,
            )
            entry["sync_exit_code"] = completed.returncode
            if completed.returncode != 0:
                entry["sync_stderr"] = completed.stderr.strip()
        dispatched.append(entry)

    state["updated_at"] = now.isoformat()
    write_state(args.state, state)
    print(json.dumps({"ok": True, "now": now.isoformat(), "dispatched": dispatched, "skipped_count": len(skipped)}, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/phantomclaw-run-due-automations.mjs
#!/usr/bin/env node

import { spawnSync } from "node:child_process";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import os from "node:os";
import path from "node:path";

const DAY_CODES = ["SU", "MO", "TU", "WE", "TH", "FR", "SA"];

function argValue(name, fallback = null) {
  const index = process.argv.indexOf(name);
  if (index === -1 || index + 1 >= process.argv.length) return fallback;
  return process.argv[index + 1];
}

function hasArg(name) {
  return process.argv.includes(name);
}

function defaultRegistryPath() {
  return path.join(os.homedir(), ".config", "phantomclaw", "automations", "registry.json");
}

function defaultStatePath() {
  return path.join(os.homedir(), ".config", "phantomclaw", "automations", "state.json");
}

function defaultOutboxDir() {
  return path.join(os.homedir(), ".config", "phantomclaw", "automation-outbox");
}

function parseRrule(rrule) {
  const parts = {};
  for (const part of String(rrule || "").split(";")) {
    const [key, ...rest] = part.split("=");
    if (!key || rest.length === 0) continue;
    parts[key.toUpperCase()] = rest.join("=");
  }
  return parts;
}

function intValues(value, fallback) {
  if (!value) return fallback;
  const values = String(value)
    .split(",")
    .map((item) => Number.parseInt(item, 10))
    .filter((item) => Number.isFinite(item));
  return values.length > 0 ? values : fallback;
}

function pad(value) {
  return String(value).padStart(2, "0");
}

function localParts(now) {
  const formatter = new Intl.DateTimeFormat("en-CA", {
    timeZone: "Europe/Berlin",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
    weekday: "short",
  });
  const parts = Object.fromEntries(formatter.formatToParts(now).map((part) => [part.type, part.value]));
  const weekdayMap = { Sun: "SU", Mon: "MO", Tue: "TU", Wed: "WE", Thu: "TH", Fri: "FR", Sat: "SA" };
  return {
    year: Number(parts.year),
    month: Number(parts.month),
    day: Number(parts.day),
    hour: Number(parts.hour),
    minute: Number(parts.minute),
    second: Number(parts.second),
    dayCode: weekdayMap[parts.weekday] ?? DAY_CODES[now.getDay()],
    date: `${parts.year}-${parts.month}-${parts.day}`,
  };
}

function dueOccurrenceKey(automation, now) {
  const parts = parseRrule(automation.rrule);
  const freq = String(parts.FREQ || "").toUpperCase();
  const local = localParts(now);
  if (parts.BYDAY) {
    const allowed = new Set(String(parts.BYDAY).split(",").map((item) => item.trim().toUpperCase()));
    if (!allowed.has(local.dayCode)) return null;
  }
  const minutes = intValues(parts.BYMINUTE, [0]);
  if (!minutes.includes(local.minute)) return null;

  if (freq === "HOURLY") {
    const interval = Math.max(1, intValues(parts.INTERVAL, [1])[0]);
    if (local.hour % interval !== 0) return null;
    return `${automation.id}:${local.date}T${pad(local.hour)}:${pad(local.minute)}`;
  }

  if (freq === "WEEKLY" || freq === "DAILY") {
    const hours = intValues(parts.BYHOUR, [0]);
    if (!hours.includes(local.hour)) return null;
    if (freq === "WEEKLY") {
      return `${automation.id}:${local.dayCode}:${local.date}:${pad(local.hour)}:${pad(local.minute)}`;
    }
    return `${automation.id}:${local.date}:${pad(local.hour)}:${pad(local.minute)}`;
  }

  return null;
}

function loadJson(filePath, fallback) {
  if (!existsSync(filePath)) return fallback;
  const text = readFileSync(filePath, "utf8").trim();
  return text ? JSON.parse(text) : fallback;
}

function slugText(value) {
  return String(value).toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "") || "automation";
}

function isoWithOffset(date) {
  return date.toISOString();
}

function runIdTimestamp(now) {
  const local = localParts(now);
  return `${local.year}${pad(local.month)}${pad(local.day)}T${pad(local.hour)}${pad(local.minute)}${pad(local.second)}+0200`;
}

function buildBundle(automation, runId, now) {
  const runner = automation.runner && typeof automation.runner === "object" ? automation.runner : {};
  const nativeRunnerAvailable = runner.status === "native" || runner.status === "native_candidate";
  const status = nativeRunnerAvailable ? "queued" : "blocked";
  const stopReason = nativeRunnerAvailable ? "ready_for_native_runner" : "native_runner_missing";
  const nowIso = isoWithOffset(now);
  const report = {
    run_id: runId,
    started_at: nowIso,
    finished_at: nowIso,
    status,
    stop_reason: stopReason,
    automation_id: automation.id,
    automation_name: automation.name,
    source_system: "codex",
    processing_system: "phantomclaw",
    codex_processing_enabled: false,
    runner_status: runner.status ?? null,
    runner_dispatch: runner.dispatch ?? null,
    rrule: automation.rrule ?? null,
    cwds: automation.cwds ?? [],
  };
  return {
    schema_version: "phantomclaw.run-bundle.v1",
    generated_at: nowIso,
    source: {
      project: "phantomclaw",
      channel: "codex-migration",
      artifact_path: automation.source?.path ?? null,
      search_url: null,
    },
    automation: {
      name: slugText(automation.id),
      label: automation.name,
      kind: "workflow",
      platform: automation.platform,
      surface: automation.surface,
      north_star_metric: null,
      parameters: {
        source_status: automation.source_status,
        runner_status: runner.status ?? null,
        rrule: automation.rrule ?? null,
      },
    },
    run: {
      run_id: runId,
      started_at: nowIso,
      finished_at: nowIso,
      status,
      stop_reason: stopReason,
      profile_name: null,
      action_events: [],
      screenshot_path: null,
    },
    metrics: {
      items_scanned: 0,
      items_considered: 0,
      actions_total: 0,
      likes_count: 0,
      reposts_count: 0,
      comments_liked_count: 0,
      follows_count: 0,
      metrics_json: {
        migration_dispatch: true,
        native_runner_available: nativeRunnerAvailable,
        codex_processing_enabled: false,
      },
    },
    report,
  };
}

const registryPath = argValue("--registry", defaultRegistryPath());
const statePath = argValue("--state", defaultStatePath());
const outboxDir = argValue("--outbox", defaultOutboxDir());
const cliPath = argValue("--phantomclaw-cli", path.join(os.homedir(), "Documents", "GitHub", "phantomclaw-cli", "dist", "cli.js"));
const sync = hasArg("--sync");
const now = argValue("--now") ? new Date(argValue("--now")) : new Date();
const registry = loadJson(registryPath, null);
if (!registry || registry.schema_version !== "phantomclaw.automation-registry.v1") {
  throw new Error(`Invalid PhantomClaw registry: ${registryPath}`);
}
const state = loadJson(statePath, { schema_version: "phantomclaw.scheduler-state.v1", last_occurrences: {} });
state.last_occurrences ||= {};

const dispatched = [];
let skippedCount = 0;
for (const automation of registry.automations || []) {
  if (automation.status !== "ACTIVE") {
    skippedCount += 1;
    continue;
  }
  const occurrence = dueOccurrenceKey(automation, now);
  if (!occurrence) {
    skippedCount += 1;
    continue;
  }
  if (state.last_occurrences[automation.id] === occurrence) {
    skippedCount += 1;
    continue;
  }
  const runId = `${automation.id}-${runIdTimestamp(now)}`;
  const bundle = buildBundle(automation, runId, now);
  const bundlePath = path.join(outboxDir, automation.id, `${runId}.bundle.json`);
  mkdirSync(path.dirname(bundlePath), { recursive: true });
  writeFileSync(bundlePath, `${JSON.stringify(bundle, null, 2)}\n`);
  state.last_occurrences[automation.id] = occurrence;
  const entry = { id: automation.id, bundle: bundlePath, occurrence };
  if (sync) {
    const result = spawnSync("node", [cliPath, "bundle", "sync", bundlePath], { encoding: "utf8" });
    entry.sync_exit_code = result.status;
    if (result.status !== 0) entry.sync_stderr = result.stderr.trim();
  }
  dispatched.push(entry);
}

state.updated_at = now.toISOString();
mkdirSync(path.dirname(statePath), { recursive: true });
writeFileSync(statePath, `${JSON.stringify(state, null, 2)}\n`);
console.log(JSON.stringify({ ok: true, now: now.toISOString(), dispatched, skipped_count: skippedCount }, null, 2));
